### ISSUE TITLE
Bad to create breaking changes when you have people using an auto-update script

### ISSUE THREAD

**Issue by pjv** [mitchellkrogza/nginx-ultimate-bad-bot-blocker#39]
when you have some people setting up and using this project, including an auto-update script that you are supplying, it seems like a bad idea to send along breaking changes through that auto-update mechanism.

i think it would be better to do one of two things:

1. never send breaking changes at all

or

2. architect an auto-update system / script that can handle things like creating new necessary .conf files so that you don't create changes that cause nginx to `[emerg]` behind people's backs.

**Comment by mitchellkrogza** [comment]
Apologies for an Nginx EMERG behind your back, recent changes just had to be made. There are constant improvements being done all the time and @itoffshore has dedicated a lot of time to getting the update and setup scripts to be universal and work on all distro's. 

The changes with all the include files I knew well in advance would cause issues for some people running auto update, it was very unfortunate but they just had to be implemented for helping users to be able to better customize the blocker to suit their own needs.

Assuring you we are doing our best to not break things behind your back.

**Comment by pjv** [comment]
Yes, I'm not saying don't make improvements. It's fantastic to make it better.

I'm saying before you make a breaking change that is definitely going to get pushed to people's production servers behind their backs (it did mine), you should set up a system so that the updates won't break their server.

It would be a lot of additional work to do. Given that you already have users out there using the auto update script, it would probably mean setting up a new version, with a new auto-update script that is capable of first updating itself before it updates any other files and then you could push whatever changes are necessary to that auto-update script to make sure that when it does the next update it won't make breaking changes (e.g., if in the last big update, the auto-update script had first update itself, then it could have been changed to include tests for the existence of the newly required `bots.d/*.conf` files and create them if they were not found.

Then you would have to keep the old system without the breaking updates running in parallel to the new version that has the self-updating auto-updates. That would keep people from breaking their production servers without knowing about it.

**Comment by pjv** [comment]
And one more thing: if you do create a breaking change, like you did, it would be nice if you could also create a script that would fix it FAST, rather than making people who might be desperate to get their production server back online re-read a long page of documentation and try to figure out which parts are new vs. which parts they already have set up.

**Comment by pjv** [comment]
And finally: sorry to be so pissy. This is a great project and I am very grateful that you put it up here and make it available for everyone. I should have said that first!

**Comment by mitchellkrogza** [comment]
@pjv once again sincere apologies, the only way to help prevent this in future is to subscribe to the mailing list. 

Also if you look in the builds section you would also quickly see notes when a major change has been introduced and what needs to be done. See the notes I added on https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/releases/tag/2017-04-20.Build-340

I am so glad the blocker is helping you, I only aim to keep making it better and no breaking changes. I am hoping the mailing list will help with keeping users notified of any upcoming major changes. But I will make sure next time there is any changes coming that they are planned for well in advance so that things like the update script etc are also updated in advance to check for the existence of all required files when pulling updates and download them automatically during the update or else the update will not proceed and email a notification saying the update did not happen because of xxxxx .... I am sure @itoffshore is still going to be able to improve greatly on the update script and I will make sure he is always informed of changes well ahead of any.

For now ..... now more major changes planned that will break the blocker .... promise 😄

**Comment by itoffshore** [comment]
@pjv - all you had to do was run:

`install-ngxblocker` (to install the new files) - this was written specifically for the big update
`setup-ngxblocker` (to insert the additional new includes) - also revised for the big update

`update-ngxblocker` is only for updating `globalblacklist.conf`- also revised with command line switches

**Comment by pjv** [comment]
@itoffshore thanks, that's good to know. Would have been better to know that before i fixed things manually. I might be blind but I didn't see those simple instructions anywhere until you just put them in here. I was not sure if the install and setup scripts were idempotent and could be safely run again on a system that has already been installed and set up.

so those scripts will not over-write an existing file like `whitelist-ips.conf` that may have been customized by the user?

**Comment by itoffshore** [comment]
@pjv it only replaces missing files so won't overwrite anything

see `check_if_updating()` in `install-ngxblocker`

I'm the package maintainer for `nginx-naxsi` in [Alpine Linux](https://pkgs.alpinelinux.org/packages?name=nginx-naxsi&branch=&repo=&arch=&maintainer=) & test everything fully on my own servers before making a `PR`.

I did put some notes in the new PR's but the `man` pages & docs could probably do with updating.

**Comment by itoffshore** [comment]
@pjv - also when you run:

`setup-ngxblocker`
`install-ngxblocker`

 **without**  `-x` the scripts print what they will update / insert (but do not make any changes).

I put these features in so by default nothing unexpected occurs.

**Comment by pjv** [comment]
@itoffshore I just tried to run the new install script on another server (where the blocker is already installed) and got this error:

`Error sourcing variables from: https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/include_filelist.txt`

This is on ubuntu 16.0.4

**Comment by itoffshore** [comment]
@pjv @mitchellkrogza - fixed for ubuntu in https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/pull/40

**Comment by itoffshore** [comment]
```
Checking url: https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/include_filelist.txt
Creating directory: /etc/nginx/conf.d
Creating directory: /etc/nginx/bots.d

REPO = https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master

Downloading [FROM]=>  [REPO]/conf.d/globalblacklist.conf            [TO]=>  /etc/nginx/conf.d/globalblacklist.conf...OK
Downloading [FROM]=>  [REPO]/conf.d/botblocker-nginx-settings.conf  [TO]=>  /etc/nginx/conf.d/botblocker-nginx-settings.conf...OK

REPO = https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master

Downloading [FROM]=>  [REPO]/bots.d/blockbots.conf              [TO]=>  /etc/nginx/bots.d/blockbots.conf...OK
Downloading [FROM]=>  [REPO]/bots.d/ddos.conf                   [TO]=>  /etc/nginx/bots.d/ddos.conf...OK
Downloading [FROM]=>  [REPO]/bots.d/whitelist-ips.conf          [TO]=>  /etc/nginx/bots.d/whitelist-ips.conf...OK
Downloading [FROM]=>  [REPO]/bots.d/whitelist-domains.conf      [TO]=>  /etc/nginx/bots.d/whitelist-domains.conf...OK
Downloading [FROM]=>  [REPO]/bots.d/blacklist-user-agents.conf  [TO]=>  /etc/nginx/bots.d/blacklist-user-agents.conf...OK
Downloading [FROM]=>  [REPO]/bots.d/blacklist-ips.conf          [TO]=>  /etc/nginx/bots.d/blacklist-ips.conf...OK
Downloading [FROM]=>  [REPO]/bots.d/bad-referrer-words.conf     [TO]=>  /etc/nginx/bots.d/bad-referrer-words.conf...OK
Downloading [FROM]=>  [REPO]/bots.d/custom-bad-referrers.conf   [TO]=>  /etc/nginx/bots.d/custom-bad-referrers.conf...OK
root@ubuntu:~# 
```

**Comment by pjv** [comment]
```
sudo ./install-ngxblocker
Checking url: https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/include_filelist.txt
Spider mode enabled. Check if remote file exists.
--2017-04-27 14:08:32--  https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/include_filelist.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.48.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.48.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 485 [text/plain]
Remote file exists.


** Dry Run ** | -x or --exec to download files

./install-ngxblocker: 132: local: /etc/nginx/bots.d: bad variable name
```

**Comment by mitchellkrogza** [comment]
Can confirm on Ubuntu 16.04

```
./install-ngxblocker

Checking url: https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/include_filelist.txt
Spider mode enabled. Check if remote file exists.
--2017-04-27 16:21:11--  https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/include_filelist.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.192.133, 151.101.64.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 485 [text/plain]
Remote file exists.


** Dry Run ** | -x or --exec to download files

./install-ngxblocker: 132: local: /etc/nginx/bots.d: bad variable name
```

and

```
./setup-ngxblocker

Checking url: https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/include_filelist.txt
Spider mode enabled. Check if remote file exists.
--2017-04-27 16:22:17--  https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/include_filelist.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 485 [text/plain]
Remote file exists.

Error sourcing variables from: https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/include_filelist.txt
```

**Comment by mitchellkrogza** [comment]
Once we get this sorted, I will try adding the testing of these scripts into the Travis tests as they run in an Ubuntu container.

**Comment by mitchellkrogza** [comment]
@itoffshore what's best ISO version of Alpine to download and run as a VM? Standard / Extended / Virtual

**Comment by itoffshore** [comment]
@mitchellkrogza - use the virtual

**Comment by mitchellkrogza** [comment]
Great thanks Stuart, never checked out Alpine so am now doing just that. Utterly amazed at the size of the ISO for Virtual 26 Mb 😮 smallest linux ISO I've ever seen yet. Going to play now with it 😀

**Comment by itoffshore** [comment]
@pjv @mitchellkrogza - change the first line of `install-ngxblocker` to `/bin/bash` (from `/bin/sh`) to get it working temporarily until I find a fix that is compatible with all platforms.

**Comment by itoffshore** [comment]
@mitchellkrogza - yes is why Alpine is the default flavor in `docker` now (& not ubuntu any more)

**Comment by mitchellkrogza** [comment]
Confirmed that https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/pull/42 fixes install-ngblocker. Thanks Stuart @itoffshore 

setup-ngxblocker still gives error

```
./setup-ngxblocker
Checking url: https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/include_filelist.txt
Spider mode enabled. Check if remote file exists.
--2017-04-27 18:13:26--  https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/include_filelist.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 485 [text/plain]
Remote file exists.

Error sourcing variables from: https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/include_filelist.txt
```

**Comment by pjv** [comment]
Confirmed also on my box that #42 ran and installed the missing .conf files to bots.d.

Thanks @itoffshore.

**Comment by mitchellkrogza** [comment]
Tested install-ngxblocker on clean Ubuntu VM ..... 100% @itoffshore 
![screen shot 2017-04-27 at 6 23 34 pm](https://cloud.githubusercontent.com/assets/9961541/25493543/e14873e6-2b76-11e7-8c02-b0a074724da2.png)

**Comment by itoffshore** [comment]
will fix `setup-ngxblocker` in the next hour or 2

**Comment by mitchellkrogza** [comment]
@itoffshore you are a STAR so appreciate your valuable time and contributions

**Comment by itoffshore** [comment]
@pjv - these scripts should hardly change now they work.

the only file that needs to change in the future is the file they source:  `include_filelist.txt`

you can find the [final files on my branch](https://github.com/itoffshore/nginx-ultimate-bad-bot-blocker/tree/alpine) to test before they are merged
